src/psyclone/tests/psyir/nodes/dynamic_omp_task_directive_test.py

```diff
@@ -0,0 +1,3915 @@
+    loops = tree.walk(Loop, stop_type=Loop)
+    loop = loops[0].children[3].children[0]
+    parent = loop.parent
+    loop.detach()
+    tdir.children[0].addchild(loop)
+    parent.addchild(tdir, index=0)
+    strans.apply(loops[0])
+    ptrans.apply(loops[0].parent.parent)
+    correct = '''subroutine my_subroutine()
+  integer, dimension(10,10) :: a
+  integer, dimension(11,10) :: b
+  integer :: i
+  integer :: j
+  integer :: k
+  integer :: l
+  integer :: m
+  integer :: n
+
+  !$omp parallel default(shared), private(i,j)
+  !$omp single
+  do i = 1, 10, 1
+    !$omp task private(j), firstprivate(l,m,n,i), shared(a,b), \
+depend(in: k,b(i + 1,:)), depend(out: a(i,:))
+    do j = l, m, n
+      a(i,j) = k
+      a(i,j) = b(i + 1,j) + k
+    enddo
+    !$omp end task
+  enddo
+  !$omp end single
+  !$omp end parallel
+
+end subroutine my_subroutine\n'''
+    assert fortran_writer(tree) == correct
+
+
+def test_omp_task_directive_37(fortran_reader, fortran_writer):
+    ''' Test the code generation correctly makes the depend clause when
+    accessing an input array shifted by the step size of the outer loop.'''
+    code = '''
+    subroutine my_subroutine()
+        type :: x
+           integer :: k
+        end type
+        integer, dimension(10, 10) :: A
+        integer, dimension(11, 10) :: B
+        integer :: i
+        integer :: j
+        integer :: k
+        type(x) :: l
+        type(x) :: m
+        type(x) :: n
+
+        do i = 1, 10, 1
+            do j = l%k, m%k, n%k
+                A(i, j) = k
+                A(i, j) = B(i+1, j) + k
+            end do
+        end do
+    end subroutine
+    '''
+    tree = fortran_reader.psyir_from_source(code)
+    ptrans = OMPParallelTrans()
+    strans = OMPSingleTrans()
+    tdir = DynamicOMPTaskDirective()
+    loops = tree.walk(Loop, stop_type=Loop)
+    loop = loops[0].children[3].children[0]
+    parent = loop.parent
+    loop.detach()
+    tdir.children[0].addchild(loop)
+    parent.addchild(tdir, index=0)
+    strans.apply(loops[0])
+    ptrans.apply(loops[0].parent.parent)
+    correct = '''subroutine my_subroutine()
+  type :: x
+    integer :: k
+  end type x
+  integer, dimension(10,10) :: a
+  integer, dimension(11,10) :: b
+  integer :: i
+  integer :: j
+  integer :: k
+  type(x) :: l
+  type(x) :: m
+  type(x) :: n
+
+  !$omp parallel default(shared), private(i,j)
+  !$omp single
+  do i = 1, 10, 1
+    !$omp task private(j), firstprivate(i), shared(l,m,n,a,b), \
+depend(in: l,m,n,k,b(i + 1,:)), depend(out: a(i,:))
+    do j = l%k, m%k, n%k
+      a(i,j) = k
+      a(i,j) = b(i + 1,j) + k
+    enddo
+    !$omp end task
+  enddo
+  !$omp end single
+  !$omp end parallel
+
+end subroutine my_subroutine\n'''
+    assert fortran_writer(tree) == correct
+
+
+def test_omp_task_directive_38(fortran_reader, fortran_writer):
+    ''' Test the code generation correctly makes the depend clause when
+    accessing an array access child of a type'''
```

Comment:
Don't we already have tests that cover this situation?

Reply:
I don't think so, I think these test specifically the case where the access is not a proxy loop variable.